src/modalities/models/coca/coca_model.py

```diff
@@ -0,0 +1,176 @@
+    ffn_hidden: Annotated[int, Field(ge=1)]
+    dropout: Annotated[float, Field(ge=0.0)]
+    bias: bool
+    attention_config: AttentionConfig
+    activation: ActivationType
+    epsilon: Annotated[float, Field(ge=0.0)]
+
+
+class CoCaConfig(BaseModel):
+    prediction_key: str = "logits"
+    vision_embd_prediction_key: str  # same key as vision encoder
+    text_embd_prediction_key: str
+    vision_cls_prediction_key: str
+    text_cls_prediction_key: str
+    vision_encoder_config: VisionTransformerConfig
+    text_decoder_config: TextDecoderConfig
+    n_pool_head: Annotated[int, Field(ge=1)]
+    n_vision_queries: Annotated[int, Field(ge=1)]
+    bias_attn_pool: bool
+    epsilon_attn_pool: Annotated[float, Field(ge=0.0)]
+    n_shared_layer: Annotated[int, Field(ge=1)]
+    weight_init: WeightInitailizationConfig
+
+
+class CoCa(NNModel):
+    """CoCa
+
+    The Contrastive Captioner (CoCa) is an encoder-decoder model that integrates the concepts of CLIP
+    and generative models such as SimVLM by using contrastive and captioning losses for training.
+
+    Paper: `CoCa: Contrastive Captioners are Image-Text Foundation Models`
+    Link: https://arxiv.org/abs/2205.01917
+    """
+
+    def __init__(
+        self,
+        prediction_key: str,
+        vision_cls_prediction_key: str,
+        text_cls_prediction_key: str,
+        vision_embd_prediction_key: str,
+        text_embd_prediction_key: str,
+        n_vision_queries: int,
+        n_pool_head: int,
+        bias_attn_pool: bool,
+        epsilon_attn_pool: float,
+        vision_encoder_config: VisionTransformerConfig,
+        text_decoder_config: GPT2Config,
+        n_shared_layer: int,
+        weight_init: WeightInitailizationConfig,
+    ) -> None:
+        super().__init__()
+        self.prediction_key = prediction_key
+        self.vision_cls_prediction_key = vision_cls_prediction_key
+        self.text_cls_prediction_key = text_cls_prediction_key
+        self.vision_embd_prediction_key = vision_embd_prediction_key
+        self.text_embd_prediction_key = text_embd_prediction_key
+
+        self.vision_encoder = VisionTransformer(**dict(vision_encoder_config))
+        self.text_decoder = TextDecoder(
+            sample_key=text_decoder_config.sample_key,
+            prediction_key=text_embd_prediction_key,
+            block_size=text_decoder_config.block_size + 1,  # +1 for the class token
+            n_layer=text_decoder_config.n_layer - n_shared_layer,
+            vocab_size=text_decoder_config.vocab_size,
+            n_head=text_decoder_config.n_head,
+            n_embd=text_decoder_config.n_embd,
+            ffn_hidden=text_decoder_config.ffn_hidden,
+            dropout=text_decoder_config.dropout,
+            bias=text_decoder_config.bias,
+            attention_config=text_decoder_config.attention_config,
+            activation=text_decoder_config.activation,
+            epsilon=text_decoder_config.epsilon,
+        )

```

Comment:
It is a bit difficult to read and understand the input arguments for `TextDecoder`. It seems that most arguments are just passed forward from `text_decoder_config`, except `text_embd_prediction_key` (2nd argument), `block_size` (3rd argument) and `n_layer` (4th argument). We should perhaps consider passing the arguments as dictionaries (essentially `text_decoder_config` attributes with a few changed key-value pairs), similar to what is done for `VisionTransformer`. A concrete option might be to use something like
```
text_decoder_config_adjusted = {
    "prediction_key": text_embd_prediction_key, 
    "block_size": text_decoder_config.block_size + 1, 
    "n_layer": text_decoder_config.n_layer - n_shared_layer
}
text_decoder_config_adjusted = dict(text_decoder_config) | text_decoder_config_adjusted
self.text_decoder = TextDecoder(**text_decoder_config_adjusted)
```
which would make the changes to `text_decode_config` more explicit and everything a bit more readable in my opinion.

Reply:
So this was basically the case before I addressed the following comment from @le1nux (See https://github.com/Modalities/modalities/pull/16/commits/d1f0369f4f19d07cb2a82d2e7717fe26126b00e5)
> instead of dumping key-value here I would use key-word arguments
> 
> ```python
> sample_key=shared_decoder_kwargs.sample_key
> prediction_key = shared_decoder_kwargs.prediction_key
> ...
> ```
> 
> The main reason is that if we `prediction_key` for instance the refactoring would not be reflected in the dictionary string keys.

I would prefer to keep the explicit keyword arguments for now.